model/task_lifecycle.go

```diff
@@ -518,14 +518,29 @@ func MarkEnd(t *task.Task, caller string, finishTime time.Time, detail *apimodel
 		return errors.Wrap(err, "marking task finished")
 	}
 
-	if err = UpdateUnblockedDependencies(t, true, "MarkEnd"); err != nil {
-		return errors.Wrap(err, "updating unblocked dependencies")
+	unblockedIds := []string{}
+	if t.IsPartOfSingleHostTaskGroup() {
+		unblockedIds, err = UpdateUnblockedDependencies(t)
```

Comment:
I don't really follow the unblocking/blocking logic in the task lifecycle, so this might be wrong - Don't tasks outside of single-host task groups still need to unblock attainable dependencies? Otherwise, how do they mark tasks that were previously unattainable as now attainable?

Reply:
The idea here is that UnblockedDependencies was added as a hot fix (that was only affecting single-host task groups) because of a bug with restarting that we didn't understand (when restarting a task, all dependencies should be marked as unattainable, so we shouldn't get into the state here that we're unblocking anything that won't be immediately re-blocked at 52).

I'm trying to understand if this is still a bug or not, since I'm not seeing any errors from our other logging that's supposed to catch this, and there've been lots of logic refactors since this was introduced. If I don't see anything suspicious, I'm going to attempt removing Unblocked altogether.

If preferred, we could make removing this under a feature flag and just handle it all together as the next piece.

Reply:
Makes sense. I guess I have a follow-on question - if unblocking is only meant for single-host task groups but the blocking is done for all tasks, where do tasks that are not in single host task groups unblock their attainable dependencies? I did some code reading, but didn't find a definitive place where unattainable would be unset when a non-single-host-task-group task finishes.

Reply:
Tasks are only set as "unattainable" when they finish, so they only become "attainable" again if they're restarted, which is something we handle in both MarkOneTaskReset and MarkTasksReset. I believe these should cover all restart cases for both single-host tasks and regular tasks -- if this logging doesn't uncover any edge cases for single-host task groups then I think we can get rid of this block completely, although possibly single-host task groups have some race with how they block old tasks.

Since this is a theory though I'll change this to still check and log everything for now, and include in the log whether or not it's a single-host task group.

Reply:
Ah yeah, it makes sense now. Thanks for the explanation!